File: lib/ui/widgets/universal_image.dart
import 'dart:io';
import 'package:cached_network_image/cached_network_image.dart';
import 'package:flutter/material.dart';
import 'package:material_symbols_icons/symbols.dart';

class UniversalImage extends StatelessWidget {
  final String? uri;
  final BoxFit? fit;
  final double? width;
  final double? height;
  final Widget? fallback;
  final IconData? fallbackIcon;
  final double? fallbackIconSize;
  final Color? fallbackIconColor;
  final BorderRadius? borderRadius;
  final bool useDecorationImage;

  const UniversalImage({
    super.key,
    this.uri,
    this.fit = BoxFit.cover,
    this.width,
    this.height,
    this.fallback,
    this.fallbackIcon = Symbols.image,
    this.fallbackIconSize = 48,
    this.fallbackIconColor = Colors.white54,
    this.borderRadius,
    this.useDecorationImage = false,
  });

  bool _isNetworkUri(String uri) {
    return uri.startsWith('http://') || uri.startsWith('https://');
  }

  Widget _buildFallback() {
    if (fallback != null) {
      return fallback!;
    }

    final icon = Icon(
      fallbackIcon,
      size: fallbackIconSize,
      color: fallbackIconColor,
    );

    if (borderRadius != null) {
      return Container(
        width: width,
        height: height,
        decoration: BoxDecoration(
          color: Colors.grey[800],
          borderRadius: borderRadius,
        ),
        child: icon,
      );
    }

    return Container(
      width: width,
      height: height,
      color: Colors.grey[800],
      child: icon,
    );
  }

  Widget _buildNetworkImage() {
    if (useDecorationImage) {
      return CachedNetworkImage(
        imageUrl: uri!,
        fit: fit,
        width: width,
        height: height,
        placeholder: (context, url) => Container(
          width: width,
          height: height,
          color: Colors.grey[800],
          child: const CircularProgressIndicator(),
        ),
        errorWidget: (context, url, error) => _buildFallback(),
      );
    }

    return CachedNetworkImage(
      imageUrl: uri!,
      fit: fit,
      width: width,
      height: height,
      placeholder: (context, url) => Container(
        width: width,
        height: height,
        color: Colors.grey[800],
        child: const CircularProgressIndicator(),
      ),
      errorWidget: (context, url, error) => _buildFallback(),
    );
  }

  Widget _buildFileImage() {
    if (useDecorationImage) {
      return Image.file(
        File(uri!),
        fit: fit,
        width: width,
        height: height,
        errorBuilder: (context, error, stackTrace) => _buildFallback(),
      );
    }

    return Image.file(
      File(uri!),
      fit: fit,
      width: width,
      height: height,
      errorBuilder: (context, error, stackTrace) => _buildFallback(),
    );
  }

  @override
  Widget build(BuildContext context) {
    if (uri == null || uri!.isEmpty) {
      return _buildFallback();
    }

    if (_isNetworkUri(uri!)) {
      return _buildNetworkImage();
    } else {
      return _buildFileImage();
    }
  }
}

File: ui/widgets/universal_image.py
"""Universal Image Widget for GroovyBox.

This module provides a UniversalImage container that handles displaying
images from various sources (local files, network URLs) with automatic
fallback to an icon when the image cannot be loaded.
"""

import flet as ft
import os
from typing import Optional
from logic.localize import tr


class UniversalImage(ft.Container):
    """An image container that supports local files and network URLs.
    
    Automatically displays a fallback icon when:
    - The URI is None or empty
    - The file doesn't exist on disk
    - The image fails to load
    
    Attributes:
        _uri: The image source URI (local path or URL).
        _fit: How the image should fit within its bounds.
        _fallback_icon: Icon to display when image is unavailable.
        _fallback_icon_size: Size of the fallback icon.
    """

    def __init__(
        self,
        uri: Optional[str] = None,
        width: Optional[float] = None,
        height: Optional[float] = None,
        border_radius: Optional[float] = None,
        fit: ft.BoxFit = ft.BoxFit.COVER,
        fallback_icon: str = ft.Icons.IMAGE,
        fallback_icon_size: float = 48,
    ):
        self._uri = uri
        self._fit = fit
        self._fallback_icon = fallback_icon
        self._fallback_icon_size = fallback_icon_size
        self._border_radius_val = border_radius

        content = self._build_content()

        super().__init__(
            content=content,
            width=width,
            height=height,
            border_radius=border_radius,
            clip_behavior=ft.ClipBehavior.ANTI_ALIAS,
            bgcolor=ft.Colors.with_opacity(0.1, ft.Colors.GREY),
        )

    def _is_network(self, uri: str) -> bool:
        """Check if a URI points to a network resource.
        
        Args:
            uri: The URI string to check.
        
        Returns:
            True if the URI starts with http:// or https://.
        """
        return uri.startswith("http://") or uri.startswith("https://")

    def _build_content(self):
        """Build the appropriate content based on the URI type.
        
        Returns:
            An Image widget for valid URIs, or a fallback Icon widget.
        """
        # Show fallback icon for missing or non-existent local files
        if not self._uri or not os.path.isfile(self._uri):
            return ft.Icon(
                icon=self._fallback_icon,
                size=self._fallback_icon_size,
                color=ft.Colors.WHITE54,
            )

        # Network or local file image
        if self._is_network(self._uri):
            return ft.Image(
                src=self._uri,
                fit=self._fit,
                error_content=ft.Icon(
                    icon=self._fallback_icon,
                    size=self._fallback_icon_size,
                    color=ft.Colors.WHITE54,
                ),
            )
        else:
            return ft.Image(
                src=self._uri,
                fit=self._fit,
                error_content=ft.Icon(
                    icon=self._fallback_icon,
                    size=self._fallback_icon_size,
                    color=ft.Colors.WHITE54,
                ),
            )
